src/core/config.ts

```diff
@@ -1,6 +1,41 @@
 import { noop, stableStringify, identity, deepEqual } from './utils'
+import {
+  ArrayQueryKey,
+  QueryKey,
+  QueryKeySerializerFunction,
+  ReactQueryConfig,
+} from './types'
 
-export const DEFAULT_CONFIG = {
+// TYPES
+
+export interface ReactQueryConfigRef {
+  current: ReactQueryConfig
+}
+
+// CONFIG
+
+export const defaultQueryKeySerializerFn: QueryKeySerializerFunction = (
+  queryKey: QueryKey
+): [string, ArrayQueryKey] => {
+  let arrayQueryKey: ArrayQueryKey = Array.isArray(queryKey)
+    ? queryKey
+    : [queryKey]
+
+  if (
+    arrayQueryKey[0] === '' ||
+    typeof arrayQueryKey[0] !== 'string' ||
+    arrayQueryKey.some(d => typeof d === 'function')
+  ) {
+    throw new Error('A valid query key is required!')
+  }
```

Comment:
We can catch bad keys like this, I think?

`JSON.stringify(key, (key, value) => { if (typeof value === 'function') { throw new Error('Bad Key') }})`